learn/building_programs/project_make.md

```diff
@@ -0,0 +1,452 @@
+    gfortran -c -o src/tests/csv_write_test.f90.o src/tests/csv_write_test.f90
+    gfortran -o src/tests/csv_write_test.exe src/tests/csv_write_test.f90.o libcsv.a
+
+There are a few things to note there, a ``make`` build usually interlaces the
+build artifacts and the source code, unless you put extra effort into implementing
+a build directory.
+Also, right now the the source files and dependencies are specified explicitly,
+which results in several additional lines even for such a simple project.
+
+
+## Automatically generated dependencies
+
+The main drawback of ``make`` for Fortran is the missing capability to
+determine module dependencies. This is usually solved by either adding those
+by hand or automatically scanning the source code with an external tool.
+Some compilers also offer to generate dependencies in ``make`` format.
+
+Before diving into the dependency generation, we will outline the concept of
+a robust take on the dependency problem.
+First, we want an approach that can process all source files independently,
+while each source file provides (``module``) or requires (``use``) modules.
+When generating the dependencies only the name of the source file and the
+module files are known, no information on the object file names should be
+required.
+
+If you check the dependency section above you will note that all dependencies are
+defined between object files rather than source files. To change this, we can
+generate a map from the source files their respective object files:
+
+```make
+# Define a map from each file name to its object file
+obj = $(src).o
+$(foreach src, $(SRCS) $(TEST_SRCS), $(eval $(src) := $(obj)))
+```
+
+Note the declaration of ``obj`` as recursively expanded variable, we effectively
+use this mechanism to define a function in ``make``. The ``foreach`` function
+allows us to loop over all source files, while the ``eval`` function allows us
+to generate ``make`` statements and evaluate them for this ``Makefile``.
+
+We adjust the dependencies accordingly as we can now define the name of the
+object files through the source file names:
+
+```make
+# Define all module interdependencies
+csv_kinds.mod := $(src/csv_kinds.f90)
+csv_module.mod := $(src/csv_module.F90)
+csv_parameters.mod := $(src/csv_parameters.f90)
+csv_utilities.mod := $(src/csv_utilities.f90)
+$(src/csv_module.F90): $(csv_utilities.mod)
+$(src/csv_module.F90): $(csv_kinds.mod)
+$(src/csv_module.F90): $(csv_parameters.mod)
+$(src/csv_parameters.f90): $(csv_kinds.mod)
+$(src/csv_utilities.f90): $(csv_kinds.mod)
+$(src/csv_utilities.f90): $(csv_parameters.mod)
+$(src/tests/csv_read_test.f90): $(csv_module.mod)
+$(src/tests/csv_test.f90): $(csv_module.mod)
+$(src/tests/csv_write_test.f90): $(csv_module.mod)
+```
+
+The same strategy of creating a map is already used for the module files, now
+it is just expanded to the object files as well.
+
+To generate the respective dependency map automatically we will use a simple
```

Comment:
```suggestion
To generate the respective dependency map automatically we will use an
```

For somebody not familiar with awk, this script is anything but simple. I'd even say it looks daunting. For such a reader it could be discouraging to read the words "simple script" and then see the thing that follows.